src/mod_csi.erl

```diff
@@ -59,17 +64,126 @@ config_spec() ->
+    case mongoose_c2s:get_mod_state(C2SData, ?MODULE) of
+        {ok, Csi} ->
+            handle_presend_element(Acc, Csi);
+        _ ->
+            {ok, Acc}
+    end.
+
+-spec handle_presend_element(mongoose_acc:t(), csi_state()) -> mongoose_c2s_hooks:result().
+handle_presend_element(Acc, Csi = #csi_state{state = inactive, buffer = Buffer, buffer_max = BMax}) ->
+    case length(Buffer) + 1 > BMax of
+        true ->
+            NewBuffer = [mark_acc_as_buffered(Acc) | Buffer],
+            NewCsi = Csi#csi_state{buffer = []},
+            ToAcc = [{state_mod, {?MODULE, NewCsi}}, {flush, lists:reverse(NewBuffer)}],
+            {stop, mongoose_c2s_acc:to_acc_many(Acc, ToAcc)};
+        _ ->
+            buffer_acc(Acc, Csi, is_acc_buffered(Acc))
+    end;
+handle_presend_element(Acc, _) ->
+    {ok, Acc}.
+
+-spec buffer_acc(mongoose_acc:t(), csi_state(), boolean()) -> mongoose_c2s_hooks:result().
+buffer_acc(Acc, Csi = #csi_state{buffer = Buffer}, false) ->
+    AccToBuffer = mark_acc_as_buffered(Acc),
+    NewCsi = Csi#csi_state{buffer = [AccToBuffer | Buffer]},
+    {stop, mongoose_c2s_acc:to_acc(Acc, state_mod, {?MODULE, NewCsi})};
+buffer_acc(Acc, _, true) ->
+    {ok, Acc}.
+
+-spec user_send_xmlel(mongoose_acc:t(), mongoose_c2s_hooks:params(), gen_hook:extra()) ->
+    mongoose_c2s_hooks:result().
+user_send_xmlel(Acc, Params, _Extra) ->
+    El = mongoose_acc:element(Acc),
+    case exml_query:attr(El, <<"xmlns">>) of
+        ?NS_CSI ->
+            {stop, handle_csi_request(Acc, Params, El)};
+        _ ->
+            {ok, Acc}
+    end.
+
+%% here we ensure CSI is active and won't buffer anymore, and the current buffer is given to sm
+-spec handle_user_stopping(mongoose_acc:t(), mongoose_c2s_hooks:params(), gen_hook:extra()) ->
+     mongoose_c2s_hooks:result().
+handle_user_stopping(Acc, #{c2s_data := C2SData}, _) ->
+    case mongoose_c2s:get_mod_state(C2SData, ?MODULE) of
+        {ok, Csi = #csi_state{buffer = Buffer}} ->
+            NewCsi = Csi#csi_state{state = active, buffer = []},
+            ToAcc = [{state_mod, {?MODULE, NewCsi}}, {flush, lists:reverse(Buffer)}],
+            {ok, mongoose_c2s_acc:to_acc_many(Acc, ToAcc)};
+        _ ->
+            {ok, Acc}
+    end.
+
+-spec reroute_unacked_messages(mongoose_acc:t(), mongoose_c2s_hooks:params(), gen_hook:extra()) ->
+    mongoose_c2s_hooks:result().
+reroute_unacked_messages(Acc, #{c2s_data := C2SData}, _) ->
+    case mongoose_c2s:get_mod_state(C2SData, ?MODULE) of
+        {ok, Csi = #csi_state{buffer = Buffer}} ->
+            mongoose_c2s:reroute_buffer(C2SData, Buffer),
+            NewCsi = Csi#csi_state{state = active, buffer = []},
+            {ok, mongoose_c2s_acc:to_acc(Acc, state_mod, {?MODULE, NewCsi})};
+        _ ->
+            {ok, Acc}
+    end.
+
+-spec handle_csi_request(mongoose_acc:t(), mongoose_c2s_hooks:params(), exml:element()) ->
+    mongoose_acc:t().
+handle_csi_request(Acc, Params, #xmlel{name = <<"inactive">>}) ->
```

Comment:
Just a note: it will crash if an element is sent with an unexpected name, right? I guess the best way would be to return a stanza error.

Reply:
True, true, added a test for it as well 👌🏽